sys/include/storage/flash_sim.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2015 Lucas Jenß
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_storage
+ * @brief
+ * @{
+ *
+ * @brief       Simulates flash memory for easier testing of storage subsystems
+ * @author      Lucas Jenß <lucas@x3ro.de>
+ */
+
+#ifndef _FS_FLASH_SIM_H
+#define _FS_FLASH_SIM_H
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Enumerates all error codes returned by the flash simulation layer
+ */
+typedef enum flash_sim_error {
+    E_FS_SUCCESS = 0,            //!< Operation was successful
+    E_FS_INVALID_WRITE = 1,      //!< Trying to set a bit from 0 -> 1 w/ a program operation
+    E_FS_INVALID_PARAMS = 2,     //!< Invalid parameters were passed to the function
+    E_FS_OUT_OF_RANGE = 3,       //!< The operation was out of range
+    E_FS_FILE_ERROR = 4,         //!< Error operating on the file simulating the flash device
+    E_FS_NOT_INITIALIZED = 5     //!< Flash simulation layer was not initialized prior to operation
+} flash_sim_error_t;
```

Comment:
Do you think this is also the case for more complex errors? When interacting with a flash device, for example, I might encounter ECC errors or a page I'm trying to read from may not have been written too and thus cannot be read. While I could represent both of those errors with `EIO`, being able to distinguish such error cases is necessary since they need to be handled differently. What would be a good solution in such a case?